src/modules/flow/update/update.c

```diff
@@ -0,0 +1,495 @@
+    mdata->base.progress_port = SOL_FLOW_NODE_TYPE_UPDATE_FETCH__OUT__PROGRESS;
+
+    return 0;
+}
+
+static void
+fetch_close(struct sol_flow_node *node, void *data)
+{
+    struct fetch_data *mdata = data;
+
+    if (mdata->base.handle)
+        sol_update_cancel(mdata->base.handle);
+
+    if (mdata->update_info)
+        sol_blob_unref(mdata->update_info);
+}
+
+static int
+cancel_fetch_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct fetch_data *mdata = data;
+
+    if (mdata->base.handle) {
+        if (!sol_update_cancel(mdata->base.handle))
+            sol_flow_send_error_packet(node, EINVAL,
+                "Could not cancel fetch process");
+        else
+            mdata->base.handle = NULL;
+    } else {
+        SOL_WRN("No current fetch in process, ignoring request to cancel");
+    }
+
+    return 0;
+}
+
+static int
+update_info_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    struct fetch_data *mdata = data;
+
+    int r;
+
+    struct sol_blob *in_value;
+
+    r = sol_flow_packet_get_json_object(packet, &in_value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    if (mdata->update_info)
+        sol_blob_unref(mdata->update_info);
+
+    mdata->update_info = sol_blob_ref(in_value);
+    SOL_NULL_CHECK(in_value, -ENOMEM);
+
+    return 0;
+}
+
+static void
+fill_update_info(struct sol_update_info *info, struct sol_blob *blob)
+{
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+
+    sol_json_scanner_init(&scanner, blob->mem, blob->size);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&scanner, &token, &key, &value, reason) {
+        if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "url"))
+            info->url = sol_json_token_get_unescaped_string_copy(&value);
+        else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "version"))
+            info->version = sol_json_token_get_unescaped_string_copy(&value);
+        else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "hash"))
+            info->hash = sol_json_token_get_unescaped_string_copy(&value);
+        else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "hash-algorithm"))
+            info->hash_algorithm = sol_json_token_get_unescaped_string_copy(&value);
+        else
+            SOL_WRN("Unknown info member: %.*s",
+                SOL_STR_SLICE_PRINT(sol_json_token_to_slice(&token)));
+    }
+}
+
+static void
+fetch_cb(void *data, int status, const char *file_path)
+{
+    struct base_data *mdata = data;
+
+    if (status < 0) {
+        sol_flow_send_error_packet(mdata->node, -status,
+            "Error while fetching update file");
+        goto end;
+    }
+
+    sol_flow_send_string_packet(mdata->node,
+        SOL_FLOW_NODE_TYPE_UPDATE_FETCH__OUT__FILE_PATH, file_path);
+end:
+    mdata->handle = NULL;
+}
+
+static void
+free_update_info(struct sol_update_info *info)
```

Comment:
call it "clear", not "free", otherwise looks like you're freeing the `info` pointer.